### PR TITLE
[MEDIA] Image croping

### PR DESCRIPTION
**Pull request by cmarianhodis**
Related issue: https://github.com/backbee/backbee-standard/issues/666

### CODE REVIEW

File: Controller/CropController.php [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+     */
+    protected $classContentRepository;
+
+    /**
+     * Crop controller's constructor
+     *
+     * @param BBApplication     $application
+     * @param RouteCollection   $routeCollection
+     */
+    public function __construct(BBApplication $application, RouteCollection $routeCollection)
```
Comment (crouillon) [inline Controller/CropController.php]: $routeCollection seems not to be used

File: Controller/CropController.php [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+     * @return boolean $response
+     * 
+     * @throws InvalidArgumentException Occurs if the provided element image is not found
+     */
+    public function cropAction()
+    {
+        $imageElement = $this->_em->find(AbstractClassContent::getClassnameByContentType('Element/Image'), $this->request->get('originalUid'));
+
+        if (null === $imageElement) {
+            throw new InvalidArgumentException('Invalid image element');
```
Comment (crouillon) [inline Controller/CropController.php]: Could you rather use SPL exception \InvalidArgumentException

File: Controller/CropController.php [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            throw new \LogicException('Element does not have a parent.');
+        }
+        // copy a new image which will be cropped
+        $newImagePathFromContent = Media::getPathFromContent($clonedImage);
+        $newImagePath = $this->application->getMediaDir().DIRECTORY_SEPARATOR.$newImagePathFromContent;
+        $oldImagePath = $this->application->getMediaDir().DIRECTORY_SEPARATOR.$clonedImage->getData('path');
+        // do the actual cropping actions persist the cloned elements
+        $this->createDirAndCopyFile($oldImagePath, $newImagePath);
+        $response = $this->doActualCrop($newImagePath);
+        $this->updateImage($clonedImage, $newImagePathFromContent, $newImagePath);
```
Comment (crouillon) [inline Controller/CropController.php]: Ther is some duplicate code here, could you please create a new private function with it ?

File: Controller/CropController.php [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+     * @param string $newImagePath  The target file path
+     *
+     * @return void
+     */
+    protected function createDirAndCopyFile($oldImagePath, $newImagePath)
+    {
+        if (!is_dir(dirname($newImagePath))) {
+            File::mkdir(dirname($newImagePath));
+        }
+        File::copy($oldImagePath, $newImagePath);
```
Comment (crouillon) [inline Controller/CropController.php]: Be aware that File::mkdir and File::copy could throw exceptions.

File: Controller/CropController.php [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+     * Do the actual crop of the image
+     *
+     * @param string $imagePath The source file path which to crop
+     * 
+     * @return boolean
+     */
+    protected function doActualCrop($imagePath)
+    {
+        CropMedia::cropImage($imagePath, $this->request->get('cropX'), $this->request->get('cropY'), $this->request->get('cropW'), $this->request->get('cropH'));
+        return Media::resize($imagePath, $imagePath, $this->request->get('cropNewW'), $this->request->get('cropNewH'));
```
Comment (crouillon) [inline Controller/CropController.php]: New line before return

Comment (crouillon) [inline Controller/CropController.php]: Is Media::resize necessary ?

File: Utils/Media.php [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+     *
+     * @return boolean $response            Returns TRUE on success, FALSE on failure
+     * 
+     * @throws BBException                  Occures if gd extension is not loaded
+     * @throws InvalidArgumentException     Occures on unsupported file type or unreadable file source
+     */
+    public static function cropImage($source, $xStartCoord, $yStartCoord, $cropWidth, $cropHeight)
+    {
+        if (false === extension_loaded('gd')) {
+            throw new BBException('gd extension is required');
```
Comment (crouillon) [inline Utils/Media.php]: Could you rather use \LogicException

File: Utils/Media.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        if (false === is_readable($source)) {
+            throw new InvalidArgumentException('Enable to read source file');
```
Comment (crouillon) [inline Utils/Media.php]: And \InvalidArgumentException

File: composer.json [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            "name": "BackBee community",
+            "homepage": "https://github.com/backbee/BackBee/graphs/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=5.4.0",
+        "backbee/backbee": "~1.1.0",
+        "backbee/bb-core-js": "~1.1.0",
+        "backbee/toolbar-bundle": "~1.1.0",
+        "backbee/utils": "~2.0"
```
Comment (crouillon) [inline composer.json]: Dependencies seems to be backbee/backbee, backbee/utils and symfony/http-foundation

**Comment by cmarianhodis** [comment]
I made the changes as per suggested

**Comment by crouillon** [comment]
Thx, @cmarianhodis :+1:

**Pull request merged**